src/StoreApi/Schemas/CartSchema.php

```diff
@@ -303,6 +303,9 @@ public function get_item_response( $cart ) {
 		// calculated so we can avoid returning costs and rates prematurely.
 		$has_calculated_shipping = $cart->show_shipping();
 
+		// Get the extended data before returning the cart - this is because the cart may be modified by extensions.
+		$extended_data = $this->get_extended_data( self::IDENTIFIER );
```

Comment:
To clarify why I'm proposing. I would run the filter here:
```PHP
apply_filters( 'woocommerce_calculated_total', $core_cart->get_total( 'total' ), $core_cart );
```

And remove this line from Subscriptions:

https://github.com/woocommerce/woocommerce-subscriptions/blob/fd2d01da50a4149e07d586a59ac73fda8e4e3cc6/includes/class-wc-subscriptions-extend-store-endpoint.php#L179

Reply:
Thanks Albert! Yes your thoughts make a lot of sense, and I agree in principle that we should try to avoid side effects when extending the API. Executing this filter is good, but we are only doing this because _we know_ (as the authors of both plugins) that the cart totals need to be recalculated here.

I am thinking about this in terms of future integrations, if the API is extended, and a new extension modifies the cart and causes a side effect (but one whose changes won't be picked up by `woocommerce_calculated_total`) the cart returned by the API won't be correct, will it? 

Moreover, if the site isn't running subscriptions, we would still be executing `apply_filters( 'woocommerce_calculated_total', $core_cart->get_total( 'total' ), $core_cart );` unnecessarily.

I think my changes leave the plugin to be as agnostic as possible about what is extending it.

Reply:
Thanks for the response @opr!

> I am thinking about this in terms of future integrations, if the API is extended, and a new extension modifies the cart and causes a side effect (but one whose changes won't be picked up by `woocommerce_calculated_total`) the cart returned by the API won't be correct, will it? 

Yes, the issue is not specific to the `woocommerce_calculated_total` filter. I was probably too fast suggesting the changes before. Instead of calling the filter directly in WC Blocks, it might be better to call a WC method which includes that filter (ie: `WC_Cart->calculate_totals()`).

I didn't dig too much into this, but my idea is that extensions would not need to run side effects inside the `get_extended_data()` callback, instead of that, WC Blocks would call the necessary filters/methods before.

> Moreover, if the site isn't running subscriptions, we would still be executing `apply_filters( 'woocommerce_calculated_total', $core_cart->get_total( 'total' ), $core_cart );` unnecessarily.

That's a good point, but the Cart and Checkout shortcodes [call `WC()->cart->calculate_totals();`](https://github.com/woocommerce/woocommerce/blob/master/includes/shortcodes/class-wc-shortcode-checkout.php#L275) (which applies `woocommerce_calculated_total`) before rendering to ensure they have the most recent values. I wonder if we should do something similar for the Rest API endpoints instead of expecting extensions to call the necessary functions to update the totals. If not, that seems something that extensions will need to account when adding support to WC Blocks because the behavior is different from shortcodes, so it might be good to document it in some way?

Reply:
I'm yet to read the full PR and convo above but I disagree about this point
```
// Get the extended data before returning the cart - this is because the cart may be modified by extensions.
		$extended_data = $this->get_extended_data( self::IDENTIFIER );
```
extended data should the last thing we fetch, all code hooking into this should get the latest and up to date cart, and the code can't have side effects, and if it has, those are going to get negletedd.
side effects can live in other filters and actions, but anything trying to extend data should only worry about shaping its data.

Reply:
>  it might be better to call a WC method which includes that filter (ie: `WC_Cart->calculate_totals()`)

I think this is a good solution, but would this remove the need for an extension to ever run side effects? I am trying to think about possible side effects that would modify the cart, and that would not be corrected by this method running... Can you think of any?

> my idea is that extensions would not need to run side effects inside the get_extended_data() callback

I think we're not worried about extensions _needing_ to run side effects, but that blocks should tolerate them when they happen, and reflect the most up to date cart. As asked above, are there any conceivable situations where the cart will be modified in such a way that `WC_Cart->calculate_totals()` won't fix it?

Reply:
I've just tested what you suggested - Adding `$cart->calculate_totals();` before we return the cart, removing the side effect from https://github.com/woocommerce/woocommerce-subscriptions/blob/fd2d01da50a4149e07d586a59ac73fda8e4e3cc6/includes/class-wc-subscriptions-extend-store-endpoint.php#L179 and moving the call to `get_extended_data` back inline. It works fine.

I do think we need to check and be sure that there are no situations we can think of where a cart can be modified by an extension, and that `$cart->calculate_totals();` won't pick this up.

Reply:
> I do think we need to check and be sure that there are no situations we can think of where a cart can be modified by an extension, and that `$cart->calculate_totals();` won't pick this up.

That's a good point. What I like about calling `$cart->calculate_totals()` is that this seems to be how the shortcodes work, so we get some confidence because it makes our implementation closer to what is already being used by extensions.

Regarding `get_extended_data()`, in my opinion we should be explicit that side effects are not allowed. Like @senadir says, WC core already has several filters/actions that can be used to modify the cart and I don't think `get_extended_data()` callbacks are the right tool for that.

There will be some cases in which the Rest API doesn't apply some core filters/actions. But I think we should evaluate those on a per-case basis: do we want to update the Rest API to support those filters or we want to provide some other way to extend it (via SlotFill, for example)?

Reply:
I'm not calling `wc()->cart->calculate_totals();` again the is best outcome here, we're already doing that in Abstract Route before we return, it seems to be an issue with Subscriptions not running its code on the right filter isteadd.